File: src/knowledge/persistence/index-persistence.ts
/**
 * IndexPersistence — 索引持久化管理器
 *
 * 职责：
 * - 将 InvertedIndex / SymbolIndex / DependencyGraph 序列化到磁盘
 * - 启动时从磁盘加载并验证文件新鲜度
 * - 检测已删除/修改/新增文件，驱动增量同步
 * - 防抖保存（避免频繁写盘）
 */

import * as fs from 'fs';
import * as path from 'path';
import * as zlib from 'zlib';
import { logger, LogCategory } from '../../logging';
import { atomicWriteFileSync } from '../../utils/atomic-write';
import { InvertedIndex, InvertedIndexSnapshot } from '../indexing/inverted-index';
import { SymbolIndex, SymbolIndexSnapshot } from '../indexing/symbol-index';
import { DependencyGraph, DependencyGraphSnapshot } from '../indexing/dependency-graph';
import { ExpansionCacheSnapshot } from '../search/query-expander';

// ============================================================================
// 类型定义
// ============================================================================

/** 持久化文件清单条目 */
interface FileManifestEntry {
  /** 最后修改时间（毫秒时间戳） */
  mtime: number;
  /** 文件大小（字节） */
  size: number;
  /** 文件类型 */
  type: 'source' | 'config' | 'doc' | 'test';
}

/** 持久化快照顶层结构 */
interface PersistenceSnapshot {
  /** 格式版本号 */
  version: number;
  /** 项目根目录 */
  projectRoot: string;
  /** 创建时间 */
  createdAt: number;
  /** 最后更新时间 */
  updatedAt: number;
  /** 文件清单（相对路径 → 元数据） */
  fileManifest: Array<[string, FileManifestEntry]>;
  /** 倒排索引快照 */
  invertedIndex: InvertedIndexSnapshot;
  /** 符号索引快照 */
  symbolIndex: SymbolIndexSnapshot;
  /** 依赖图快照 */
  dependencyGraph: DependencyGraphSnapshot;
  /** 查询扩展缓存快照（可选，向后兼容） */
  expansionCache?: ExpansionCacheSnapshot;
}

/** 新鲜度验证结果 */
export interface FreshnessResult {
  /** 未变化的文件 */
  unchanged: string[];
  /** 已修改的文件 */
  modified: string[];
  /** 已删除的文件（索引中有，磁盘上无） */
  deleted: string[];
  /** 新增的文件（磁盘上有，索引中无） */
  added: string[];
}

/** 当前格式版本 */
const PERSISTENCE_VERSION = 1;

/** 默认防抖延迟（毫秒） */
const DEFAULT_DEBOUNCE_MS = 5000;

// ============================================================================
// IndexPersistence 类
// ============================================================================

export class IndexPersistence {
  private cacheFilePath: string;
  /** 旧版未压缩路径（用于迁移清理） */
  private legacyCacheFilePath: string;
  private debounceTimer: ReturnType<typeof setTimeout> | null = null;
  private debounceMs: number;

  constructor(projectRoot: string, debounceMs = DEFAULT_DEBOUNCE_MS) {
    this.cacheFilePath = path.join(projectRoot, '.magi', 'cache', 'search-index.json.gz');
    this.legacyCacheFilePath = path.join(projectRoot, '.magi', 'cache', 'search-index.json');
    this.debounceMs = debounceMs;
  }

  /**
   * 保存索引到磁盘
   */
  save(
    projectRoot: string,
    invertedIndex: InvertedIndex,
    symbolIndex: SymbolIndex,
    dependencyGraph: DependencyGraph,
    files: Array<{ path: string; type: 'source' | 'config' | 'doc' | 'test' }>,
    expansionCache?: ExpansionCacheSnapshot
  ): void {
    try {
      const dir = path.dirname(this.cacheFilePath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }

      // 构建文件清单
      const fileManifest: Array<[string, FileManifestEntry]> = [];
      for (const file of files) {
        try {
          const fullPath = path.join(projectRoot, file.path);
          const stat = fs.statSync(fullPath);
          fileManifest.push([file.path, {
            mtime: stat.mtimeMs,
            size: stat.size,
            type: file.type,
          }]);
        } catch {
          // 文件可能已被删除，跳过
        }
      }

      const snapshot: PersistenceSnapshot = {
        version: PERSISTENCE_VERSION,
        projectRoot,
        createdAt: Date.now(),
        updatedAt: Date.now(),
        fileManifest,
        invertedIndex: invertedIndex.toJSON(),
        symbolIndex: symbolIndex.toJSON(),
        dependencyGraph: dependencyGraph.toJSON(),
        expansionCache,
      };

      const jsonStr = JSON.stringify(snapshot);
      const compressed = zlib.gzipSync(jsonStr);
      atomicWriteFileSync(this.cacheFilePath, compressed);

      // 清理旧版未压缩文件
      if (fs.existsSync(this.legacyCacheFilePath)) {
        try { fs.unlinkSync(this.legacyCacheFilePath); } catch { /* 忽略 */ }
      }

      logger.info('索引持久化.保存成功', {
        files: fileManifest.length,
        rawSize: `${(jsonStr.length / 1024).toFixed(0)}KB`,
        compressedSize: `${(compressed.length / 1024).toFixed(0)}KB`,
        path: this.cacheFilePath,
      }, LogCategory.SESSION);
    } catch (error) {
      logger.warn('索引持久化.保存失败', { error, path: this.cacheFilePath }, LogCategory.SESSION);
    }
  }

  /**
   * 防抖保存：延迟写盘，合并多次连续调用
   */
  debouncedSave(
    projectRoot: string,
    invertedIndex: InvertedIndex,
    symbolIndex: SymbolIndex,
    dependencyGraph: DependencyGraph,
    files: Array<{ path: string; type: 'source' | 'config' | 'doc' | 'test' }>,
    expansionCache?: ExpansionCacheSnapshot
  ): void {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }
    this.debounceTimer = setTimeout(() => {
      this.debounceTimer = null;
      this.save(projectRoot, invertedIndex, symbolIndex, dependencyGraph, files, expansionCache);
    }, this.debounceMs);
  }

  /**
   * 从磁盘加载索引快照
   * 返回 null 表示无可用缓存
   */
  load(): PersistenceSnapshot | null {
    try {
      let raw: string;

      if (fs.existsSync(this.cacheFilePath)) {
        // 优先读取压缩格式
        const compressed = fs.readFileSync(this.cacheFilePath);
        raw = zlib.gunzipSync(compressed).toString('utf-8');
      } else if (fs.existsSync(this.legacyCacheFilePath)) {
        // 向后兼容：读取旧版未压缩格式
        raw = fs.readFileSync(this.legacyCacheFilePath, 'utf-8');
      } else {
        return null;
      }

      const snapshot = JSON.parse(raw) as PersistenceSnapshot;

      // 版本校验
      if (snapshot.version !== PERSISTENCE_VERSION) {
        logger.info('索引持久化.版本不匹配，丢弃缓存', {
          cached: snapshot.version,
          current: PERSISTENCE_VERSION,
        }, LogCategory.SESSION);
        return null;
      }

      logger.info('索引持久化.加载成功', {
        files: snapshot.fileManifest.length,
        updatedAt: new Date(snapshot.updatedAt).toISOString(),
      }, LogCategory.SESSION);

      return snapshot;
    } catch (error) {
      logger.warn('索引持久化.加载失败', { error }, LogCategory.SESSION);
      return null;
    }
  }

  /**
   * 验证文件新鲜度：对比持久化清单与当前文件系统
   */
  validateFreshness(
    projectRoot: string,
    snapshot: PersistenceSnapshot,
    currentFiles: Array<{ path: string; type: 'source' | 'config' | 'doc' | 'test' }>
  ): FreshnessResult {
    const result: FreshnessResult = {
      unchanged: [],
      modified: [],
      deleted: [],
      added: [],
    };

    const manifest = new Map(snapshot.fileManifest);
    const currentFileSet = new Set(currentFiles.map(f => f.path));

    // 检查清单中的文件：是否删除或修改
    for (const [filePath, entry] of manifest) {
      if (!currentFileSet.has(filePath)) {
        result.deleted.push(filePath);
        continue;
      }

      try {
        const fullPath = path.join(projectRoot, filePath);
        if (!fs.existsSync(fullPath)) {
          result.deleted.push(filePath);
          continue;
        }
        const stat = fs.statSync(fullPath);
        if (Math.abs(stat.mtimeMs - entry.mtime) > 1) {
          // mtime 有变化 → 文件已修改
          result.modified.push(filePath);
        } else {
          result.unchanged.push(filePath);
        }
      } catch {
        result.deleted.push(filePath);
      }
    }

    // 检查新增文件：当前存在但清单中没有
    for (const file of currentFiles) {
      if (!manifest.has(file.path)) {
        result.added.push(file.path);
      }
    }

    return result;
  }

  /**
   * 恢复索引并执行增量同步
   * 返回 { restored: true, expansionCache } 表示成功恢复，{ restored: false } 表示需要全量重建
   */
  restoreAndSync(
    projectRoot: string,
    invertedIndex: InvertedIndex,
    symbolIndex: SymbolIndex,
    dependencyGraph: DependencyGraph,
    currentFiles: Array<{ path: string; type: 'source' | 'config' | 'doc' | 'test' }>
  ): { restored: boolean; expansionCache?: ExpansionCacheSnapshot } {
    const snapshot = this.load();
    if (!snapshot) return { restored: false };

    // 验证新鲜度
    const freshness = this.validateFreshness(projectRoot, snapshot, currentFiles);

    const changeCount = freshness.modified.length + freshness.deleted.length + freshness.added.length;

    // 如果变化文件超过总文件数的 30%，直接全量重建更高效
    const totalFiles = snapshot.fileManifest.length + freshness.added.length;
    if (totalFiles > 0 && changeCount / totalFiles > 0.3) {
      logger.info('索引持久化.变化过多，全量重建', {
        changeCount,
        totalFiles,
        ratio: `${Math.round(changeCount / totalFiles * 100)}%`,
      }, LogCategory.SESSION);
      return { restored: false };
    }

    // 从快照恢复索引
    try {
      invertedIndex.fromJSON(snapshot.invertedIndex);
      symbolIndex.fromJSON(snapshot.symbolIndex);

      // DependencyGraph 需要当前文件集合
      const fileSet = new Set(currentFiles.map(f => f.path));
      dependencyGraph.fromJSON(snapshot.dependencyGraph, projectRoot, fileSet);
    } catch (error) {
      logger.warn('索引持久化.恢复失败，需全量重建', { error }, LogCategory.SESSION);
      return { restored: false };
    }

    // 执行增量同步
    // 1. 删除已不存在的文件
    for (const filePath of freshness.deleted) {
      invertedIndex.removeFile(filePath);
      symbolIndex.removeFile(filePath);
      dependencyGraph.removeFile(filePath);
    }

    // 2. 更新已修改的文件
    for (const filePath of freshness.modified) {
      invertedIndex.updateFile(projectRoot, filePath);
      symbolIndex.updateFile(projectRoot, filePath);
      dependencyGraph.updateFile(projectRoot, filePath);
    }

    // 3. 新增文件
    for (const filePath of freshness.added) {
      invertedIndex.updateFile(projectRoot, filePath);
      symbolIndex.updateFile(projectRoot, filePath);
      dependencyGraph.updateFile(projectRoot, filePath);
    }

    logger.info('索引持久化.增量同步完成', {
      unchanged: freshness.unchanged.length,
      modified: freshness.modified.length,
      deleted: freshness.deleted.length,
      added: freshness.added.length,
    }, LogCategory.SESSION);

    return { restored: true, expansionCache: snapshot.expansionCache };
  }

  /**
   * 删除缓存文件
   */
  invalidate(): void {
    try {
      for (const filePath of [this.cacheFilePath, this.legacyCacheFilePath]) {
        if (fs.existsSync(filePath)) {
          fs.unlinkSync(filePath);
        }
      }
      logger.info('索引持久化.缓存已删除', undefined, LogCategory.SESSION);
    } catch {
      // 忽略删除失败
    }
  }

  /**
   * 取消待执行的防抖保存
   */
  dispose(): void {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
      this.debounceTimer = null;
    }
  }
}

File: crates/magi-knowledge-store/src/index_persistence.rs
use std::collections::HashMap;
use std::fs;
use std::io::{Read as IoRead, Write as IoWrite};
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::dependency_graph::DependencyGraphSnapshot;
use crate::inverted_index::InvertedIndexSnapshot;
use crate::query_expander::ExpansionCacheSnapshot;
use crate::symbol_index::SymbolIndexSnapshot;

const PERSISTENCE_VERSION: u32 = 1;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FileManifestEntry {
    pub mtime: u64,
    pub size: u64,
    pub file_type: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PersistenceSnapshot {
    pub version: u32,
    pub project_root: String,
    pub created_at: u64,
    pub updated_at: u64,
    pub file_manifest: Vec<(String, FileManifestEntry)>,
    pub inverted_index: InvertedIndexSnapshot,
    pub symbol_index: SymbolIndexSnapshot,
    pub dependency_graph: DependencyGraphSnapshot,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expansion_cache: Option<ExpansionCacheSnapshot>,
}

#[derive(Clone, Debug, Default)]
pub struct FreshnessResult {
    pub unchanged: Vec<String>,
    pub modified: Vec<String>,
    pub deleted: Vec<String>,
    pub added: Vec<String>,
}

impl FreshnessResult {
    pub fn change_count(&self) -> usize {
        self.modified.len() + self.deleted.len() + self.added.len()
    }
}

pub struct IndexPersistence {
    cache_file_path: PathBuf,
    legacy_cache_file_path: PathBuf,
}

impl IndexPersistence {
    pub fn new(project_root: &str) -> Self {
        let base = Path::new(project_root).join(".magi").join("cache");
        Self {
            cache_file_path: base.join("search-index.json.gz"),
            legacy_cache_file_path: base.join("search-index.json"),
        }
    }

    pub fn save(&self, snapshot: &PersistenceSnapshot) -> Result<(), String> {
        let dir = self
            .cache_file_path
            .parent()
            .ok_or("invalid cache path")?;
        fs::create_dir_all(dir).map_err(|e| format!("mkdir failed: {}", e))?;

        let json_str =
            serde_json::to_string(snapshot).map_err(|e| format!("serialize failed: {}", e))?;

        let compressed = compress_gzip(json_str.as_bytes())?;

        fs::write(&self.cache_file_path, &compressed)
            .map_err(|e| format!("write failed: {}", e))?;

        if self.legacy_cache_file_path.exists() {
            let _ = fs::remove_file(&self.legacy_cache_file_path);
        }

        Ok(())
    }

    pub fn load(&self) -> Option<PersistenceSnapshot> {
        let raw = if self.cache_file_path.exists() {
            let compressed = fs::read(&self.cache_file_path).ok()?;
            decompress_gzip(&compressed).ok()?
        } else if self.legacy_cache_file_path.exists() {
            fs::read(&self.legacy_cache_file_path).ok()?
        } else {
            return None;
        };

        let snapshot: PersistenceSnapshot = serde_json::from_slice(&raw).ok()?;

        if snapshot.version != PERSISTENCE_VERSION {
            return None;
        }

        Some(snapshot)
    }

    pub fn validate_freshness(
        &self,
        project_root: &str,
        snapshot: &PersistenceSnapshot,
        current_files: &[(String, String)],
    ) -> FreshnessResult {
        let mut result = FreshnessResult::default();

        let manifest: HashMap<&str, &FileManifestEntry> = snapshot
            .file_manifest
            .iter()
            .map(|(k, v)| (k.as_str(), v))
            .collect();

        let current_set: std::collections::HashSet<&str> =
            current_files.iter().map(|(p, _)| p.as_str()).collect();

        for (file_path, entry) in &manifest {
            if !current_set.contains(file_path) {
                result.deleted.push(file_path.to_string());
                continue;
            }

            let full_path = Path::new(project_root).join(file_path);
            match fs::metadata(&full_path) {
                Ok(meta) => {
                    let mtime = meta
                        .modified()
                        .ok()
                        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                        .map(|d| d.as_millis() as u64)
                        .unwrap_or(0);

                    if (mtime as i64 - entry.mtime as i64).unsigned_abs() > 1 {
                        result.modified.push(file_path.to_string());
                    } else {
                        result.unchanged.push(file_path.to_string());
                    }
                }
                Err(_) => {
                    result.deleted.push(file_path.to_string());
                }
            }
        }

        for (file_path, _) in current_files {
            if !manifest.contains_key(file_path.as_str()) {
                result.added.push(file_path.clone());
            }
        }

        result
    }

    pub fn should_full_rebuild(freshness: &FreshnessResult, total_files: usize) -> bool {
        if total_files == 0 {
            return true;
        }
        let change_count = freshness.change_count();
        change_count as f64 / total_files as f64 > 0.3
    }

    pub fn build_file_manifest(
        project_root: &str,
        files: &[(String, String)],
    ) -> Vec<(String, FileManifestEntry)> {
        let mut manifest = Vec::with_capacity(files.len());
        for (path, file_type) in files {
            let full_path = Path::new(project_root).join(path);
            if let Ok(meta) = fs::metadata(&full_path) {
                let mtime = meta
                    .modified()
                    .ok()
                    .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                    .map(|d| d.as_millis() as u64)
                    .unwrap_or(0);
                manifest.push((
                    path.clone(),
                    FileManifestEntry {
                        mtime,
                        size: meta.len(),
                        file_type: file_type.clone(),
                    },
                ));
            }
        }
        manifest
    }

    pub fn invalidate(&self) {
        for path in [&self.cache_file_path, &self.legacy_cache_file_path] {
            if path.exists() {
                let _ = fs::remove_file(path);
            }
        }
    }
}

fn compress_gzip(data: &[u8]) -> Result<Vec<u8>, String> {
    use flate2::write::GzEncoder;
    use flate2::Compression;

    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder
        .write_all(data)
        .map_err(|e| format!("gzip compress: {}", e))?;
    encoder
        .finish()
        .map_err(|e| format!("gzip finish: {}", e))
}

fn decompress_gzip(data: &[u8]) -> Result<Vec<u8>, String> {
    use flate2::read::GzDecoder;

    let mut decoder = GzDecoder::new(data);
    let mut output = Vec::new();
    decoder
        .read_to_end(&mut output)
        .map_err(|e| format!("gzip decompress: {}", e))?;
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gzip_roundtrip() {
        let original = b"hello world, this is test data for gzip compression";
        let compressed = compress_gzip(original).unwrap();
        assert!(compressed.len() > 0);
        let decompressed = decompress_gzip(&compressed).unwrap();
        assert_eq!(decompressed, original);
    }

    #[test]
    fn test_freshness_result() {
        let mut f = FreshnessResult::default();
        f.modified.push("a.rs".into());
        f.deleted.push("b.rs".into());
        f.added.push("c.rs".into());
        assert_eq!(f.change_count(), 3);
    }

    #[test]
    fn test_should_full_rebuild() {
        let mut f = FreshnessResult::default();
        for i in 0..4 {
            f.modified.push(format!("{}.rs", i));
        }
        assert!(IndexPersistence::should_full_rebuild(&f, 10));
        f.modified.clear();
        f.modified.push("x.rs".into());
        assert!(!IndexPersistence::should_full_rebuild(&f, 10));
    }

    #[test]
    fn test_load_nonexistent() {
        let persistence = IndexPersistence::new("/tmp/nonexistent_project_12345");
        assert!(persistence.load().is_none());
    }
}
